app/src/behaviors/behavior_tap_dance.c

```diff
@@ -0,0 +1,340 @@
+    tap_dance->release_at = event.timestamp + tap_dance->config->tapping_term_ms;
+    int32_t ms_left = tap_dance->release_at - k_uptime_get();
+    LOG_DBG("%d tap-dance ms_left = %d", tap_dance->position, ms_left);
+    if (ms_left > 0) {
+        k_delayed_work_submit(&tap_dance->release_timer, K_MSEC(ms_left));
+    }
+}
+
+// press_tap_dance_behavior
+
+// Returns behavior_keymap_binding_pressed to ensure that keycode is actually pressed
+static inline int press_tap_dance_behavior(struct active_tap_dance *tap_dance, int64_t timestamp) {
+    // Press chosen tap dance binding
+    struct zmk_behavior_binding binding =
+        tap_dance->config
+            ->behaviors[MIN(tap_dance->counter, tap_dance->config->behavior_count) - 1];
+    struct zmk_behavior_binding_event event = {
+        .position = tap_dance->position,
+        .timestamp = timestamp,
+    };
+    return behavior_keymap_binding_pressed(&binding, event);
+}
+
+// release_tap_dance_behavior
+
+// Returns behavior_keymap_binding_pressed to ensure that keycode is actually released
+static inline int release_tap_dance_behavior(struct active_tap_dance *tap_dance,
+                                             int64_t timestamp) {
+    // Release chosen tap dance binding
+    struct zmk_behavior_binding binding =
+        tap_dance->config
+            ->behaviors[MIN(tap_dance->counter, tap_dance->config->behavior_count) - 1];
+    struct zmk_behavior_binding_event event = {
+        .position = tap_dance->position,
+        .timestamp = timestamp,
+    };
+    clear_tap_dance(tap_dance);
+    return behavior_keymap_binding_released(&binding, event);
+}
+
+// on_tap_dance_binding_pressed
+
+// Code that determines what happens when a tap-dance binding is pressed.
+// Does not register a keycode press unless the tap-dance's counter is
+// equal to the maximum number of bindings associated with it.
+// E.g. if a defined tap-dance has bindings = <&kp N1>, <&kp N2>, <&kp N3>;
+// when the counter is equal to 3, press_tap_dance_behavior will be called
+// to register N3's keycode being pressed
+static int on_tap_dance_binding_pressed(struct zmk_behavior_binding *binding,
+                                        struct zmk_behavior_binding_event event) {
+    // Code executes when the tap dance key is pressed.
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_tap_dance_config *cfg = dev->config;
+    struct active_tap_dance *tap_dance;
+    tap_dance = find_tap_dance(event.position);
+    // No active tap dance found on keypress location, attempt to create a new one
+    if (tap_dance == NULL) {
+        if (new_tap_dance(event.position, cfg, &tap_dance) == -ENOMEM) {
+            LOG_ERR("Unable to create new tap dance. Insufficient space in active_tap_dances[].");
+            return ZMK_BEHAVIOR_OPAQUE;
+        }
+        // Tap dance successfully created. Reset timer.
```

Comment:
I would remove this entire block of code and the return, so the same flow is taken for both the first key press and any consecutive presses.

```
        // Tap dance successfully created. Reset timer.
        LOG_DBG("%d created new tap dance", event.position);
        reset_timer(tap_dance, event);
        return ZMK_BEHAVIOR_OPAQUE;
```

(I think there is a bug in the current implementation of the tap-dance - tap_dance->is_pressed is not set to true on the first keypress, causing the key to be released when the timer expires instead of on release).

Reply:
Actually there is a bug - Tap dances with 1 behavior will never trigger the behavior keypress because we're returning early here.

Reply:
Confirming that the single-keypress bug is resolved in updated squashed commit 22118f5